### PR TITLE
Telemetry definitions and Elixir support

### PR DESCRIPTION
**Pull request by bryannaegele**
Definitions support has been added along with Elixir support to make things simple for those users. A breaking change introduced is moving `spannable_events` to return a list of tuples rather than a map. When implementing the otel bridge, it was apparent that a map isn't useful and `list_events` already returns a tuple list.

**Outstanding Questions**

1. I'm not convinced that we should support plain events. There's a strong case to be made that users should be forced to define events completely at declaration time. On the other hand, it ties the hands of users creating convenience span macros a bit. I'm inclined to keep the support for only taking an event name but not documenting it as a middle ground.
2. Storage is currently using persistent term. Otel is OTP 21+. Should we move back to ets to extend support? I'm inclined to not and wait for user feedback. If there's demand, it's not a big deal. On the other hand, given the usage there is no real benefit gained by storing it as a term.

### CODE REVIEW

File: .gitignore
Comment (josevalim) [inline .gitignore]: FWIW, this should go in your global .gitignore. :)

File: lib/telemetry_registry.ex
Comment (josevalim) [inline lib/telemetry_registry.ex]: I don't think you need to define and store the docs upfront. When someone first calls `format_telemetry_docs`, you can traverse all events, build the docs, and return them. :)

File: lib/telemetry_registry.ex
Comment (josevalim) [inline lib/telemetry_registry.ex]: You need to escape the interpolation here OR use the `~S"""` sigil to disable them. :)

```suggestion
  \#{format_telemetry_docs(@telemetry_docs)}
```

File: lib/telemetry_registry.ex
Comment (josevalim) [inline lib/telemetry_registry.ex]: Besides the comment above about not storing `@telemetry_docs` upfront, I would make this a macro so we can hide the module attributes (i.e. they should be implementation details). Something like this:

```elixir
defmacro format_telemetry_docs do
  quote do
    TelemetryRegistry.__docs__(__MODULE__)
  end
end

@doc false
def __docs__(module) do
  Module.get_attribute(module, :telemetry_event)
  |> Enum.map(&format_event_to_doc/1)
  |> IO.iodata_to_binary()
end
```

WDYT?

File: lib/telemetry_registry.ex
Comment (hauleth) [inline lib/telemetry_registry.ex:163]: Are we sure we want such macro? For me it feels quite unidiomatic.

File: lib/telemetry_registry.ex
Comment (hauleth) [reply lib/telemetry_registry.ex]: @josevalim module attributes cannot be implementation detail, as it would make Erlang implementation impossible from what I see. So I would say, that the module attributes will be **the** implementation.

File: lib/telemetry_registry.ex
Comment (hauleth) [inline lib/telemetry_registry.ex]: This assumes that docs will always be the Markdown. I would say that this isn't something that library like Telemetry should assume.

Comment (josevalim) [reply lib/telemetry_registry.ex]: It is totally ok for the Elixir `telemetry_registry` bindings to assume this. If you have something else, then you simply don't call this function. :)

File: lib/telemetry_registry.ex
Comment (josevalim) [reply lib/telemetry_registry.ex]: I am ok with with using `@telemetry_event` but the `@telemetry_docs` one should not be necessary/exposed.

FWIW, there are advantages to the macro approach, such as validating arguments, being able to provide defaults, etc. Given you need to `use TelemetryRegistry` module, I don't see the macro being a huge issue, BUT I can also see the appeal on keeping it close to the Erlang counter-part.

File: lib/telemetry_registry.ex
Comment (hauleth) [reply lib/telemetry_registry.ex]: I think, that there should be no docs by default, and instead the documentation generator should take care about it.

Comment (josevalim) [reply lib/telemetry_registry.ex]: We talked about this in the last meeting. ExDoc is not going to hardcode Telemetry into it and an extensible mechanism is likely years away, since the focus of the ExDoc team is on Erlang integration. So it is either this or nothing for the foreseeable feature. And I would hate having to declare my events in three places (on execute, as an attribute, and as docs).

File: .gitignore
Comment (bryannaegele) [reply .gitignore]: I blame Tristan. I copy/pastad from the otel libs 😆

File: lib/telemetry_registry.ex
Comment (bryannaegele) [reply lib/telemetry_registry.ex]: As noted, we reached this conclusion last week. If we need to change it down the road to add other formatters, it's just an implementation detail to make it configurable and leave markdown as the default.

File: lib/telemetry_registry.ex
Comment (bryannaegele) [reply lib/telemetry_registry.ex]: I was copying from how ex_unit tests get rolled up. I prefer not to add a secondary attribute, but hadn't considered another macro to do it. I like it. 👍

File: lib/telemetry_registry.ex
Comment (bryannaegele) [reply lib/telemetry_registry.ex:163]: I'm torn to be honest. If I don't preload the docs, it isn't necessary. However, it also enables adding validation for Elixir users (noted that I haven't added that yet). This could be extended to Erlang users via an Erlang macro, though I'm not sure they'd use it.

It's not necessary that they use this macro. The only "requirement" is adding the `use TelemetryRegistry` to set things up for using the attributes.

File: lib/telemetry_registry.ex
Comment (davydog187) [inline lib/telemetry_registry.ex]: Should `TelemetryRegistry` expose a formatter export for `telemetry_event` for no parens? 

```suggestion
    telemetry_event %{
      event: [:test_elixir_app, :single, :event],
      description: "emitted when this event happens",
      measurements: "%{duration: non_neg_integer()}",
      metadata: "%{status: status(), name: String.t()}"
    }
```

File: lib/telemetry_registry.ex
Comment (hauleth) [reply lib/telemetry_registry.ex]: @josevalim, it still can be done in the function/macro that will generate the documentation text instead of there. IMHO that would make it clearer IMHO.

By the way, I was thinking about either adding the support for extensions to ExDoc or writing my own documentation generator that would use ASCIIDoc with support for extensions so it would be possible to generate documentation for other libraries that use metadata (like `OpenApiSpex`).

Comment (josevalim) [reply lib/telemetry_registry.ex]: >  it still can be done in the function/macro that will generate the documentation text instead of there.

Do you mean something like: `TelemetryRegistry.docs_for(SomeExternalModule)`? If so, I actually like your proposal too, as it decouples the definition from the docs.

> By the way, I was thinking about either adding the support for extensions to ExDoc 

Right - I don't doubt it is possible. The big question is when it is going to happen. Are we ok with postponing this feature indefinitely in favor of something that may or may not exist?

Comment (hauleth) [reply lib/telemetry_registry.ex]: For now I would go with moving documentation generation to `format_telemetry_docs/0` macro. So instead of using `telemetry_event` macro which generates default docs, generate them only when such macro will be used.

Comment (josevalim) [reply lib/telemetry_registry.ex]: Right, that's what i originally proposed. But now I think the latest API I proposed is better because it can be used to generate documentation for the current module and external modules.

It would be something like this:

    def docs_for(module) do
      get_events(module) |> format_docs()
    end

where 

    defp get_events(module) do
      Module.get_attribute(module, :telemetry_event)
    rescue
      _ -> module.__info__(:attributes)[:telemetry_event]
    end

Or something like that. So something like Phoenix can list `use TelemetryRegistry` and register the events on the modules that dispatch them but still have unified docs in a separate module.

Comment (hauleth) [reply lib/telemetry_registry.ex]: Why not:

```elixir
defp get_events(module) do
  {:docs_v1, _, _, _, _, metadata, _} = Code.fetch_docs(module)

  Map.get(metadata, :telemetry_events, [])
end
```

Which would also create base for future ExDoc integration.

File: lib/telemetry_registry.ex
Comment (josevalim) [reply lib/telemetry_registry.ex]: :+1:

File: lib/telemetry_registry.ex
Comment (josevalim) [reply lib/telemetry_registry.ex]: The above doesn't work if the module is being defined and we already store them in the attributes anyway, so I don't see what it would buy us. :)

**Comment by bryannaegele** [comment]
@josevalim and @hauleth I've attempted to coalesce all of the great feedback into a nicer API. Let me know what you think. Also, please comment the outstanding questions noted in https://github.com/beam-telemetry/telemetry_registry/pull/5#issue-445891590 😁

File: lib/telemetry_registry.ex
Comment (bryannaegele) [reply lib/telemetry_registry.ex:163]: Leaving this as a macro for future use (validation and such). Users are free to declare attributes by hand.

File: lib/telemetry_registry.ex
Comment (josevalim) [inline lib/telemetry_registry.ex]: Protip! Use `~S'''` in the parent @doc/@moduledoc so in here you can use `@moduledoc """` without escaping!

File: lib/telemetry_registry.ex
Comment (josevalim) [inline lib/telemetry_registry.ex:183]: What do you think about my later suggestion of actually passing the module? And we try to load the available events either internally (module attribute) or externally?

File: lib/telemetry_registry.ex
Comment (josevalim) [inline lib/telemetry_registry.ex:176]: I think we can kill this macro now, no? Or you want to keep it for convenience?

**Comment by josevalim** [comment]
Looks great! I dropped just two tiny comments.

File: lib/telemetry_registry.ex
Comment (hauleth) [inline lib/telemetry_registry.ex]:
```suggestion
        |> Enum.map(fn
          [event] when is_map(event) -> event
          event -> event
        end)
```

Comment (josevalim) [reply lib/telemetry_registry.ex]: Good call. It can also be a `List.flatten` call.

Comment (bryannaegele) [reply lib/telemetry_registry.ex]: I think flatten won't work because it will flatten multiple plain events to one list of atoms, e.g. `List.flatten([[:a, :b, :c], [:d, :e, :f], [%{}])` results in `[:a, :b, :c, :d, :e, :f, %{}]`

**Comment by bryannaegele** [comment]
I think I'm going to add support for the definition values for `measurements` and `metadata` to accept a keyword list for anyone that wants to go all out like https://github.com/keathley/finch/blob/master/lib/finch/telemetry.ex. WDYT?

File: lib/telemetry_registry.ex
Comment (bryannaegele) [reply lib/telemetry_registry.ex:176]: It seems like a nice convenience to keep.

**Comment by bryannaegele** [comment]
I have removed support for plain event declaration since that defeats the agenda of getting events properly documented for users. 

I would like to add support for enhanced definitions similar to https://github.com/keathley/finch/blob/master/lib/finch/telemetry.ex by accepting a proplist of `[{atom, binary}]` in place of a string and formatting accordingly, but that can be a feature addition at a later date. I'd like to get this release out to allow release of the otel bridge library in the near future.

Resolves #4

File: lib/telemetry_registry.ex
Comment (josevalim) [inline lib/telemetry_registry.ex:48]: Do we need to have Erlang examples in the Elixir docs? 🤔 Maybe this should be moved to the README?

**Review by josevalim — APPROVED** [review]
Looks great! My only comment is about how the docs should be structured so it works for both Erlang and Elixir, but it is your call. :)

**Review by hauleth — APPROVED** [review]

File: lib/telemetry_registry.ex
Comment (hauleth) [reply lib/telemetry_registry.ex:48]: As there will be only one docs published, and there is no support for Erlang docs in ExDocs then I think it would be nice to have such information there, at least for now, until EEP48 fully lands.

Comment (bryannaegele) [reply lib/telemetry_registry.ex:48]: I'll defer to @hauleth on this. What's the timeframe on that? This year?

**Pull request merged**
